Fixes the "wrong" computation of the AP score for newer sklearn versions.

Pandoro opened the pull request:
Given that sklearn changed the way AP scores are computed this implements a custom version.
This implementation follows the official Market-1501 computation of the AP.

Pandoro committed:
Fixes the "wrong" computation of the AP score.

Given that sklearn changed the way AP scores are computed this implements a custom version.
This implementation follows the official Market-1501 computation of the AP.

lucasb-eyer commented on README.md:
```diff
@@ -273,6 +273,9 @@ The evaluation code in this repository simply uses the scikit-learn code, and th
 Other frameworks have [the same problem](https://github.com/Cysu/open-reid/issues/50), but we expect many not to be aware of this.
 
+To make the evaluating code independent of the sklearn version we have implemented our own version of the average precision computation.
+This now follows the official Market1501 code and results in values directly comparable.
```
Suggested change (diff bold):
> This now follows the official Market1501 code **independently of the installed sklearn version** and **always** results in values directly comparable.

Pandoro replied:
I didn't go with that suggestions, since honestly I didn't see it, but as discussed earlier, the default behavior is now as it always was, but it now suggests to use the flag in order to be consistent.

lucasb-eyer commented on evaluate.py:
```diff
@@ -52,6 +51,43 @@
+                         'got lengths y_true:{} and y_score:{}'.format(
+                            len(y_true), len(y_score)))
+
+    y_true_sorted = y_true[np.argsort(-y_score, kind='mergesort')]
```
Maybe add comment that it's set to `mergesort` because we want stable sorting? Otherwise, who knows, a future commit might remove it thinking it's only for speed or something.

Pandoro replied:
Done!

lucasb-eyer reviewed (commented):
Also suggest changing the PR title to append "for newer sklearn versions"

Code looks correct to me and I bet you compared it with known values.

My actual concern is if we really want to do this? We already mention in the readme how to get either version of the score, and no paper besides ours mentions whether they actually use the Market1501 code or their own? At least one other framework which people actively use doesn't.

Alternatively, we could make it a flag?

Pandoro committed:
Changed the new implementation of the "old" eval to be a flag.

As discussed in the pull request, using the default market evaluation is now an optional flag.
This also adds a comment about why we use mergesort and updates the README according to the new changes.

Pandoro commented:
@lucasb-eyer please have another quick look at this and tell me if you are happy with the naming and the changes, or if there is other feedback so that we can maybe merge this soon.
